### PR TITLE
Pull review/1414

### PR DESCRIPTION
**Pull request by hvacengi**
PR with progress so far on for my review.

### CODE REVIEW

File: doc/source/structures/collections/list.rst [hunk 1]
```diff
@@ -41,6 +41,9 @@ Structure
           - Type
           - Description
 
+        * - All suffixes of :struct:`Enumerable`
+          -
+          - :struct:`List` objects are a type of :struct:`Enumerable`
```
Comment (Dunbaratu) [inline doc/source/structures/collections/list.rst:46]: Man I should really look into whether or not Sphinx has a sort of include operator that can be used for cases like this.  We have a lot of them.

File: doc/source/structures/collections/range.rst [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    This is one of the few places in kOS where there is a distinction between
+    decimal (floating point, or fractional) scalar numbers and whole (integer,
+    or round) scalar numbers.  Using a decimal scalar will not throw an error,
+    however it may give unexpected results due to rounding.
+
+There are 3 ways of constructing a :struct:`Range`:
+
+- :code:`RANGE(FROM, TO, STEP)`
+  Will create a sequence of numbers that starts with `FROM` and contains all numbers `N + STEP` such that `N` belongs to the sequence and that are smaller than `TO`. For example
+  :code:`RANGE(3,8,2)` will contain numbers 3, 5 and 7. If `FROM` > `TO` then the sequence will be descending. For example :code:`RANGE(2,-9,3)` will contain numbers 2, -1, -4 and -7.
```
Comment (Dunbaratu) [inline doc/source/structures/collections/range.rst:19]: I'm not sure how I feel about this.  I didn't look at tomek's range feature in detail until now and I'm not sure I like it not including the endpoint and yet using the terminology "from...to".

"Count from 1 to 10" doesn't mean to stop at 9.  Even when a mathematical concept uses [0..n-1] rather than [1..n], the way you usually speak that aloud is "zero to N minus one", not "zero to N".

I get the mathematical reason to not include the endpoint, but maybe the terminology should be a bit different somehow.  Calling it "TO" really gives the wrong message.  You're not counting TO this value, but counting to just shy of it.  Maybe call them "begin" and "end"?

File: doc/source/structures/collections/range.rst [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+    This is one of the few places in kOS where there is a distinction between
+    decimal (floating point, or fractional) scalar numbers and whole (integer,
+    or round) scalar numbers.  Using a decimal scalar will not throw an error,
+    however it may give unexpected results due to rounding.
+
+There are 3 ways of constructing a :struct:`Range`:
+
+- :code:`RANGE(FROM, TO, STEP)`
+  Will create a sequence of numbers that starts with `FROM` and contains all numbers `N + STEP` such that `N` belongs to the sequence and that are smaller than `TO`. For example
```
Comment (Dunbaratu) [inline doc/source/structures/collections/range.rst:18]: The phrase "that are smaller than" is wrong when counting backward with a negative STEP.  Then the endpoint is at the low end of the range, not the high end.

File: doc/source/structures/collections/list.rst (hunk 1 above)
Comment (Dunbaratu) [inline doc/source/structures/collections/list.rst:46]: Technically, it's a kind of `Collection` now, and a `Collection`, in turn, is a kind of `Enumerable`.  Thus the 'CLEAR' suffix is part of `Collection`.  But I'm not sure we want this to jump through two levels of link before we get to LENGTH, CONTAINS, And so on.

(the same issue arises with the other collections - queue, etc).

File: doc/source/structures/collections/list.rst [hunk 4]
```diff
@@ -50,33 +53,18 @@ Structure
         * - :meth:`REMOVE(index)`
           - None
           - remove item by index
-        * - :attr:`CLEAR`
+        * - :meth:`CLEAR()`
```
Comment (Dunbaratu) [inline doc/source/structures/collections/list.rst:56]: Technically this clear suffix is part of `Collection` now.  But I don't know if we want to have that much extra clicking to get to the `Enumerable` suffixes (see comment above).

(the same issue arises with the other collections - queue, etc).

File: src/kOS.Safe/Encapsulation/EnumerableValue.cs [hunk 5]
```diff
@@ -30,23 +34,44 @@ IEnumerator IEnumerable.GetEnumerator()
...
-            AddSuffix("ITERATOR",   new NoArgsSuffix<Enumerator>          (() => new Enumerator(Collection.GetEnumerator())));
-            AddSuffix("CONTAINS",   new OneArgsSuffix<BooleanValue, T>    (item => Collection.Contains(item)));
-            AddSuffix("EMPTY",      new NoArgsSuffix<BooleanValue>        (() => !Collection.Any()));
-            AddSuffix("DUMP",       new NoArgsSuffix<StringValue>         (() => new StringValue(ToString())));
+            AddSuffix("ITERATOR",        new NoArgsSuffix<Enumerator>(() => new Enumerator(InnerEnumerable.GetEnumerator())));
+            AddSuffix("REVERSEITERATOR", new NoArgsSuffix<Enumerator>(() => new Enumerator(Enumerable.Reverse(InnerEnumerable).GetEnumerator())));
+            AddSuffix("LENGTH",          new NoArgsSuffix<ScalarValue>(() => InnerEnumerable.Count()));
+            AddSuffix("CONTAINS",        new OneArgsSuffix<BooleanValue, T>((n) => Contains(n)));
+            AddSuffix("EMPTY",           new NoArgsSuffix<BooleanValue>(() => !InnerEnumerable.Any()));
+            AddSuffix("DUMP",            new NoArgsSuffix<StringValue>(() => new StringValue(ToString())));
```
Comment (Dunbaratu) [inline src/kOS.Safe/Encapsulation/EnumerableValue.cs:74]: I prefer the consistency in the deleted formatting here.  Either you prefer having columns over having tight spacing, or you prefer having tight spacing over having columns, but this is a weird half and half mixture.  You removed the spaces that made column 3 line up, but kept the ones that made column 2 line up.

File: src/kOS.Safe/Encapsulation/EnumerableValue.cs [hunk 6]
```diff
@@ -1,26 +1,30 @@
...
     {
-        protected TC Collection { get; private set; }
+        protected TE InnerEnumerable { get; private set; }
+        private readonly string label;
 
-        protected EnumerableValue(TC collection)
+        protected EnumerableValue(string label, TE enumerable)
         {
-            Collection = collection;
+            this.label = label;
```
Comment (Dunbaratu) [inline src/kOS.Safe/Encapsulation/EnumerableValue.cs:19]: hmm.  What might be better is to pass in the GetType() of the derived type rather than a string label.  Then set the string label using KOSNomenclature.GetKOSName().  Thus we guarantee we are printing a message to the user using the type names that the rest of kOS is showing them for the type, and we future-proof it against future decisions to change that kOS type name.

File: src/kOS.Safe/Encapsulation/RangeValue.cs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+using System;
+using System.Collections.Generic;
+
+namespace kOS.Safe
+{
+    [kOS.Safe.Utilities.KOSNomenclature("Range")]
+    public class RangeValue : EnumerableValue<ScalarIntValue, Range>
+    {
+        private const string DumpFrom = "from";
+        private const string DumpTo = "to";
```
Comment (Dunbaratu) [inline src/kOS.Safe/Encapsulation/RangeValue.cs:14]: See comment above in the .rst doc about it being strange to use the words "from" and "to" to describe a range with an open endpoint that doesn't actually go all the way TO the "to" value.

File: src/kOS.Safe/Encapsulation/QueueValue.cs [hunk 8]
```diff
@@ -13,71 +13,59 @@ public QueueValue() : this(new Queue<T>())
...
-            Collection.Enqueue(val);
+            InnerEnumerable.Enqueue(val);
         }
 
         public override Dump Dump()
         {
             var result = new DumpWithHeader
             {
-                Header = "QUEUE of " + Collection.Count() + " items:"
+                Header = "QUEUE of " + InnerEnumerable.Count() + " items:"
```
Comment (Dunbaratu) [inline src/kOS.Safe/Encapsulation/QueueValue.cs:35]: Hmm. Maybe get the name from KOSNomenclature.GetKOSName(GetType()) instead of hardcoding the word "QUEUE" here, thus putting all the configuration of what we call this thing when talking to the user into one single place - the attribute.  That same thing would apply to all the other places we hardcode the type names like this.

Come to think of it, we could add that as a get property to the base class `Structure`, so all Structure's can just use KOSName as a property.

File: src/kOS/Module/Bootstrapper.cs [hunk 9]
```diff
@@ -24,6 +24,7 @@ public void Start()
 
             CheckForLegacyArchive();
 
+            KOSNomenclature.PopulateMapping(typeof(kOS.Safe.Encapsulation.Structure).Assembly, this.GetType().Assembly);
```
Comment (Dunbaratu) [inline src/kOS/Module/Bootstrapper.cs:27]: Doing it from here is definitely an improvement.

File: doc/source/structures/collections/lexicon.rst [hunk 10]
```diff
@@ -46,7 +46,7 @@ Constructing a lexicon
 ----------------------
 
 If you wish to make your own lexicon from scratch you can do so with the
-LEXICON() built-in function::
+LEXICON() or LEX() built-in function::
```
Comment (Dunbaratu) [inline doc/source/structures/collections/lexicon.rst:49]: I think we should add a "LEX" alias KOSNomenclature for Lexicon.cs to go with this.  See GeoCoordinates.cs as an example of the way that can be done.

File: doc/source/structures/collections/list.rst (hunk 1 above)
Comment (tomekpiotrowski) [inline doc/source/structures/collections/list.rst:46]: I wanted to avoid an extra page just for a single method.

File: doc/source/structures/collections/range.rst (hunk 2 above)
Comment (tomekpiotrowski) [inline doc/source/structures/collections/range.rst:19]: Feel free to change the naming. Python uses start/stop for example.

File: doc/source/structures/collections/range.rst (hunk 3 above)
Comment (tomekpiotrowski) [inline doc/source/structures/collections/range.rst:18]: I thought that things will be clear by looking at the example below, but sure, we can add an extra sentence explaining that case.

File: src/kOS.Safe/Encapsulation/EnumerableValue.cs (hunk 5 above)
Comment (tomekpiotrowski) [inline src/kOS.Safe/Encapsulation/EnumerableValue.cs:74]: Yeah, normally I never do those kind of indents, but I'm trying to maintain them when seeing one in the code. Apparently I'm not very good at this ;P

File: src/kOS.Safe/Encapsulation/EnumerableValue.cs (hunk 6 above)
Comment (tomekpiotrowski) [inline src/kOS.Safe/Encapsulation/EnumerableValue.cs:19]: :+1:

File: doc/source/structures/collections/range.rst (hunk 2 above)
Comment (hvacengi) [inline doc/source/structures/collections/range.rst:19]: I was actually going to discuss the same thing, until I started looking and found that it's the same limitation of python.  And because the default is to start at 0 this method allows you to work with zero based indices:

```
set l to list("one", "two", "three").
for i in range(l:length) {
    print l[i] + " (" + i + ")".
}
```

I'm not sure that changing the closing parameter to "end" will really convey that it is an "exclusive" limit any better than "to".  Like so many things, it's difficult to convey the mathematical meaning in a single word.  I suppose for the purpose of documentation the name of the parameter is only useful for helping the KS users understand the meaning, so we could simple change it to "from_inclusive" and "to_exclusive" (though that's implying an affinity towards "_" deliminated naming)

File: doc/source/structures/collections/range.rst (hunk 3 above)
Comment (hvacengi) [inline doc/source/structures/collections/range.rst:18]: Oh, I forgot I was planning on making a revision to that line in my original review.  I was going to rephrase it to "belongs to the sequence and is between `FROM` and `TO` (including `FROM`)"

File: doc/source/structures/collections/list.rst (hunk 4 above)
Comment (hvacengi) [inline doc/source/structures/collections/list.rst:56]: Lists are the only thing that inherit from CollectionValue actually (currently).  Queue and Stack inherit directly from EnumerableValue.  So List is really the only structure where `Clear` shows up.

File: src/kOS.Safe/Encapsulation/EnumerableValue.cs (hunk 5 above)
Comment (hvacengi) [inline src/kOS.Safe/Encapsulation/EnumerableValue.cs:74]: Actually this was my change.  I was attempting to compromise.  I do not like the extra white space  that comes with the indentation, but others think it helps to clarify.  So I thought that it would end up being OK if the methods themselves weren't indented, but the constructor was.  I'd rather just eliminate the indentation outright.

File: src/kOS.Safe/Encapsulation/EnumerableValue.cs (hunk 6 above)
Comment (hvacengi) [inline src/kOS.Safe/Encapsulation/EnumerableValue.cs:19]: Yeah, I couldn't make those types of changes in the initial review of Tom's PRs because I didn't have Nomenclature.  I intend to go through and find this kind of thing if possible.  I also want to make Argument Mismatch exceptions show the new Nomenclature names.

File: src/kOS.Safe/Encapsulation/QueueValue.cs (hunk 8 above)
Comment (hvacengi) [inline src/kOS.Safe/Encapsulation/QueueValue.cs:35]: Adding that method to Structure is a great idea.  If we make the change here, we'll want to make sure that it's handled throughout the serialization methods.

File: doc/source/structures/collections/lexicon.rst (hunk 10 above)
Comment (hvacengi) [inline doc/source/structures/collections/lexicon.rst:49]: Good catch.

File: doc/source/structures/collections/range.rst (hunk 2 above)
Comment (Dunbaratu) [inline doc/source/structures/collections/range.rst:19]: I think this is a case where ambiguity would actually be an improvement.  "Start at 1, Stop at 10" could mean either [1..9] or [1..10] depending on interpretation, or "Begin at 1, End at 10" could mean either [1..9] or [1..10] depending on interpretation, but "From 1 to 10" only means [1..10].  It's an unambiguous meaning, but unfortunately its unambiguously the wrong meaning.  An ambiguous meaning that has to be specified in the docs would be an improvement over an unambiguous one that has to be contradicted in the docs.

File: src/kOS.Safe/Encapsulation/EnumerableValue.cs (hunk 6 above)
Comment (Dunbaratu) [inline src/kOS.Safe/Encapsulation/EnumerableValue.cs:19]: Excellent idea about the argument mismatch exceptions.

File: src/kOS.Safe/Encapsulation/EnumerableValue.cs (hunk 5 above)
Comment (Dunbaratu) [inline src/kOS.Safe/Encapsulation/EnumerableValue.cs:74]: Whatever your opinion on the clarifying (ahem, not "extra") spaces are, the original way is consistent with how the other InitializeSuffixes() throughout the mod are doing it.  I get my nice formatting overridden in this mod so much that the one time there actually does seem to be a practice in place that goes the other way and does it the way I like, I'm going to stick up for it.

File: doc/source/structures/collections/list.rst (hunk 4 above)
Comment (Dunbaratu) [inline doc/source/structures/collections/list.rst:56]: What's the rationale for not being able to `CLEAR` a queue or a stack?

File: doc/source/structures/collections/list.rst (hunk 1 above)
Comment (Dunbaratu) [inline doc/source/structures/collections/list.rst:46]: Okay I can see that.  It's just that when people do `PRINT LIST(1,2,3):INHERITANCE.` they'll see the Collection type in the printout that's not documented, between List and Enumerable.  Maybe that's not such a big deal?

File: src/kOS.Safe/Encapsulation/RangeValue.cs [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+            if (step < 1)
+            {
+                throw new KOSException("Step must be a positive integer");
+            }
+        }
+
+        private void InitializeRangeSuffixes()
+        {
+            AddSuffix("FROM", new NoArgsSuffix<ScalarValue>(() => InnerEnumerable.From));
+            AddSuffix("TO", new NoArgsSuffix<ScalarValue>(() => InnerEnumerable.To));
```
Comment (Dunbaratu) [inline src/kOS.Safe/Encapsulation/RangeValue.cs:51]: Ah crud - From/To the terminology isn't just in the docs.  It's also exposed here.  That means it bugs me a bit more than it otherwise would.  Docs can be changed or argued about post-release more easily than what the suffix is actually called can be.  Once we release with the exposed name in the suffix then we have to keep it that way or break scripts.
